hangman/lib/hangman/game.ex

```diff
@@ -0,0 +1,133 @@
+defmodule Hangman.Game do
+
+  defstruct game_state: :initializing,   # :won | :lost | :already_used | :good_guess | :bad_guess | :initializing
+            turns_left: 7,               # the number of turns left (game starts with 7)
+            letters: [],                 # a list of single character strings. If a letter in a particular
+                                         # position has been guessed, that letter will appear in `letters`. 
+                                         # Otherwise, it will be shown as an underscore
+            used: [],                    # A sorted list of the letters already guessed
+            last_guess: "",              # the last letter guessed by the player
+            word: ""                     # the secret word attempting to be guessed
+
+  def new_game() do
+    word = Dictionary.random_word
+    %Hangman.Game{
+      word: word,
+      letters: get_underscore_list(word)
+    }
+  end
+
+  def tally(game) do
+    game
+    |> Map.from_struct
+    |> Map.delete(:word)
+  end
+
+  def make_move(state = %{word: word, used: used, letters: letters}, guess) do
+    game = state 
+    |> Map.put(:last_guess, guess)
+
+    letter_in_word = word |> String.contains?(guess)
+    letter_is_used = guess in used
+    letters_unguessed = "_" in letters
+    valid_letter = String.match?(guess, ~r/^[a-z]$/)
+
+    do_guess(game, letter_in_word, letter_is_used, letters_unguessed, valid_letter)
+  end
```

Comment:
This is interesting. You've basically precalculated every factor that might decide the result, and then selected the response in one call.

It wasn't what I was expecting, simply because it makes for a longer function, and (as you discovered) it's hard to decode the various calls to `do_guess`.

My initial reaction was do deduct something just because of all the local variables and the length of the function. But, at the same time, I like the fact that it's basically a recipe card selection algorithm from the 1960's :)

So, no deduction, but a strong suggestion.

Rather than computing everything, is there a decision tree that would let you calculate just one or two things initially, and then calculate further details as needed. This would lead to (perhaps) two or three small functions rather than one big one.